addons/Rakugo/lib/systems/Parser.gd

```diff
@@ -109,220 +109,257 @@ func _init():
 	var ret := 0
-	
+
 	if s[0] == '	':
 		for i in s.length():
 			var c = s[i]
-		
+
 			if c == '	':
 				ret += 1
 			else:
 				break
-	
+
 	return ret
 
 func remove_double_quotes(s:String) -> String:
 	return s.substr(1, s.length()-2)
 
 func do_parse_script(file_name:String):
 	var file = File.new()
-	
+
 	if file.open(file_name, File.READ) != OK:
 		prints("Parser", "can't open file : " + file_name)
 		return
-	
-	var indent_count:int
-	
-	var menu_choices
-	
-	var current_menu_result
-	
+
+	var indent_count: int
+	var menu_choices = []
+	var current_menu_result: RegExMatch
+
 	while !stop_thread:
 		if state == State.Menu and file.eof_reached():
+			# the previous line was the last of a list of options in a menu
 			state = State.Normal
-				
-#			prints("Parser", "parse_script", "mod Normal")
-			
 			if !menu_choices.empty():
 				parse_array.push_back(["MENU", current_menu_result, menu_choices])
+				menu_choices = []
 			else:
 				break
 		elif file.eof_reached():
 			break
-		
-		var line = file.get_line()
 
+		var line = file.get_line()
 		if line.empty():
 			continue
 
-		#erase tabulations
 		# TODO handle indentation levels
 		indent_count = count_indent(line)
-		
+		# erase tabulations
 		line = line.lstrip('	')
-
 		if state == State.Menu and indent_count == 0:
+			# the previous line was the last of a list of options in a menu
 			state = State.Normal
-				
-#			prints("Parser", "parse_script", "mod Normal")
-			
 			if !menu_choices.empty():
+				# push the built menu into the parsed tree
 				parse_array.push_back(["MENU", current_menu_result, menu_choices])
+				menu_choices = []
 
-		match(state):
-			State.Normal:
-				for key in regex_cache:
-					var result = regex_cache[key].search(line)
-					
-					if result:
-						match(key):
-							"MENU":
-								current_menu_result = result
-					
-								menu_choices = []
-								
-								state = State.Menu
-
-								labels[result.get_string("label")] = parse_array.size()
-								
-							"DIALOGUE":
-								var dialogue_label = result.get_string("label")
-				
-								labels[dialogue_label] = parse_array.size()
-							
-							_:
-								parse_array.push_back([key, result])
-			State.Menu:
-				var result = regex_cache["CHOICE"].search(line)
-				if result:
-#					prints("Parser", "parse_script", "CHOICE")
-#
-#					for key in result.names:
-#						prints(" ", key, result.get_string(key))
-						
-					menu_choices.push_back(result)
-					
-					continue
+		state = _parse_single_line(line, state, menu_choices, current_menu_result)
 
 	file.close()
-	
+
 	prints("Parser", "do_parse_script", "end")
 
+
+func _parse_single_line(
+	line: String, current_state: int, current_menu_options: Array, current_menu_result: RegExMatch
+) -> int:
+	assert(current_state in State.values(), "The function argument is expected to be a State value")
+	var new_state = current_state
+	match (current_state):
+		State.Normal:
+			for key in regex_cache:
```

Comment:
Question here: even after matching for something on the line, we're still trying to match something else rather than going to the next line. Is this right?

Reply:
Woups... I forgot a break, I made an Issue ! Thanks :)